src/metaswitch/clearwater/queue_manager/etcd_synchronizer.py

```diff
@@ -0,0 +1,253 @@
+                
+            executor.shutdown()    
+
+        _log.info("Quitting FSM")
+        self._fsm.quit()
+
+    def fsm_timer_expired(self):
+        self._timer_flag = True;
+
+    def wait_for_fsm(self):
+        while not self._timer_flag and not self._terminate_flag:
+            sleep(self.WAIT_FOR_TIMER_POP)
+
+    def fsm_loop(self, etcd_value=None): # Change to add helper message
+        queue_config = {}
+
+        try:
+            if etcd_value is None:
+                if self._last_value is None: # pragma: no cover
+                    queue_config = json.loads(self.default_value())
+                else:
+                    queue_config = json.loads(self._last_value)
+            else:
+                queue_config = json.loads(etcd_value)
+        except: # pragma: no cover
+            queue_config = json.loads(self.default_value())
+
+        self._fsm.fsm_update(queue_config)
+        etcd_updated_value = json.dumps(queue_config)
+
+        # If we have a new state, try and write it to etcd.
+        if etcd_updated_value != self._last_value:
+            _log.debug("Writing updated queue config to etcd")
+
+            if not self.write_to_etcd(etcd_updated_value):
+                self.fsm_loop(etcd_updated_value)
+
+    # Write the new cluster view to etcd. We may be expecting to create the key
+    # for the first time.
+    def write_to_etcd(self, queue_config):
+        _log.debug("Writing state {} into etcd with index {}"
+                   .format(queue_config, self._index))
+        needs_no_retry = True
+
+        try:
+            if self._index:
+                self._client.write(self.key(), queue_config, prevIndex=self._index)
+            else: # pragma: no cover
+                self._client.write(self.key(), queue_config, prevExist=False)
+        except (EtcdAlreadyExist, ValueError): # pragma: no cover
+            _log.debug("Contention on etcd write")
+            # Our etcd write failed because someone got there before us. We
+            # don't need to retry in this case as we'll just pick up the
+            # changes in the next etcd read
+        except Exception as e:
+            # Catch-all error handler (for invalid requests, timeouts, etc) -
+            # unset our state and start over.
+            _log.error("{} caught {!r} when trying to write {} with index {}"
+                       " - pause before retrying"
+                       .format(self._ip, e, queue_config, self._index))
+
+            # Setting last_cluster_view to None means that the next successful
+            # read from etcd will trigger the state machine, which will mean
+            # that any necessary work/state changes get retried.
+            self._last_value, self._last_index = None, None
+
+            # Sleep briefly to avoid hammering a failed server
+            self.pause()
+            needs_no_retry = False
+     
+        return needs_no_retry
+
+    def remove_from_queue(self, successful):
```

Comment:
I think this function and add_to_queue (and maybe other places I haven't got to yet?) are unnecessarily complex, because they manipulate this reasonably complicated JSON document inline. I think if you pulled that responsibility out into a separate QueueConfig class, which exposed methods to answer the questions this code needed to ask ("is this node the first entry?", "is this node currently restarting?") everything would become a lot clearer and simpler.